rct_optimizations/src/rct_optimizations/dh_chain.cpp

```diff
@@ -5,98 +5,60 @@ namespace rct_optimizations
 {
 // DH Transform
 
-DHTransform::DHTransform(std::array<double, 4> params_, DHJointType type_)
-  : params(std::move(params_))
+DHTransform::DHTransform(const Eigen::Vector4d& params_, DHJointType type_)
+  : params(params_)
   , type(type_)
 {
 }
 
-Eigen::Isometry3d DHTransform::createRelativeTransform(const double joint_value) const
-{
-  Eigen::Isometry3d transform(Eigen::Isometry3d::Identity());
-
-  // Create aliases for the d and theta parameters such that the joint value could be added to them appropriately
-  double d = params[0];
-  double theta = params[1];
-  double r = params[2];
-  double alpha = params[3];
-  switch (type)
-  {
-  case DHJointType::LINEAR:
-    d += joint_value;
-    break;
-  case DHJointType::REVOLUTE:
-    theta += joint_value;
-    break;
-  default:
-    break;
-  }
-
-  // Perform the DH transformations
-  transform.translate(Eigen::Vector3d(0.0, 0.0, d));
-  transform.rotate(Eigen::AngleAxisd(theta, Eigen::Vector3d::UnitZ()));
-  transform.translate(Eigen::Vector3d(r, 0.0, 0.0));
-  transform.rotate(Eigen::AngleAxisd(alpha, Eigen::Vector3d::UnitX()));
-
-  return transform;
-}
-
 double DHTransform::createRandomJointValue() const
 {
   std::mt19937 mt_rand = std::mt19937(std::random_device{}());
   std::uniform_real_distribution<double> dist(min, max);
```

Comment:
I did a little more reading on `std::mt19937` and the recommended way to set it up is to have just one generator as a class member (`mt_rand` in this case) and use it for all the random number generation that the class instance has to do. This is because it's more computationally-expensive to seed and initialize the generator than to generate a new random number.

However, getting a new random number increments the generator so `createRandomJointValue` couldn't be `const` anymore.

Reply:
Is it worth removing the `const` specifier?

Reply:
I would prefer keeping the functions `const` because it indicates that calling them won't change the current state of the class which is generally nice to know. I could make `mt_rand` a global variable in that source file, but that doesn't seem like a great practice. If you think `std::mt19937` has issues generating random enough numbers when we are continuously creating new objects, I can change it. If it's just an efficiency thing, then I might be inclined to leave it since this method is only called when generating simulated observations for a chain

Reply:
Technically using a class-member `mt19937` generator to generate a new random number _does_ change the state of the class since the `mt19937` generator gets updated every time it's called. I wish there was some way to indicate the scope of a non-`const` function, since generating random numbers doesn't touch the stored transform info or anything else important.

It's probably not a big deal since in the grand scheme of things we're only generating a few random numbers, but if we ever need to make thousands and thousands then not re-initializing a generator every time may make a difference in the time needed to process. Seems like something we can leave as-is for the time being.

Reply:
Maybe using a static `mt19937` in that function is the right way to go. It would allow us to keep the functions const but not have to create a new instance every time the function is called.

Reply:
I think that would be OK.